src/module/ovs/bpf/kernel_upcall_tp.bpf.c

```diff
@@ -0,0 +1,46 @@
+#include <openvswitch.h>
+#include <bpf/bpf_core_read.h>
+
+#include <common.h>
+#include "ovs_common.h"
+
+/* Please keep in sync with its Rust counterpart in crate::module::ovs::bpf.rs. */
+struct upcall_event {
+	u8 cmd;
+	u32 port;
+	u32 cpu;
+} __attribute__((packed));
+
+/* Hook for raw_tracepoint:openvswitch:ovs_dp_upcall. */
+DEFINE_HOOK(
+	struct dp_upcall_info *upcall;
+	struct upcall_context uctx = {};
+	struct upcall_event *upcall_event;
+	u64 tid = bpf_get_current_pid_tgid();
+
+	upcall = (struct dp_upcall_info *) ctx->regs.reg[3];
```

Comment:
So, you mean extending the `ProbeOffsets` struct for every argument use by every hook in every module?

I was assuming we'd use that for widely-used / common arguments but for module-specific ones like this one it seemed to me like we'd cause this structure to explode unnecessarily. 

Maybe we can add some "custom" offsets? Something like:
```rust

pub(super) struct ProbeOffsets {
    pub(super) sk_buff: i8,
    pub(super) skb_drop_reason: i8,
    pub(super) net_device: i8,
    pub(super) net: i8,
    pub(super) custom: [i8; MAX_CUSTOM],
}
``` 

Then we could open the door for setting the custom offset from the module:

```rust
let mut upcall_tp = Probe::raw_tracepoint(Symbol::from_name("openvswitch:ovs_dp_upcall")?)?;
upcall_tp.set_custom_offset(0, "struct dp_upcall_info *");
...
```

Reply:
I was thinking this mechanism would only be needed for module-specific hooks so it's only one module giving meaning to the custom stuff.
For offsets, would we be able to use a map of maps to store all available arguments so a hook would just have to access offset[ksym]["struct whatever *"]?

Reply:
> > I was thinking this mechanism would only be needed for module-specific hooks so it's only one module giving meaning to the custom stuff.
> 
> Yes but even target hooks can use a shared probe.
>

Right, but generic hooks use generic arguments, not custom ones right?
 
> > For offsets, would we be able to use a map of maps to store all available arguments so a hook would just have to access offset[ksym]["struct whatever *"]?
> 
> But we can't use strings as a key and not doing so would mean to statically define keys in the core, which is what we where trying to avoid.

Can't we use strings as keys? why not?

Reply:
not char*, but char[MAX_ARG_LEN] should work IIUC;

Reply:
> No I mean we can have two targeted hooks from two different modules using the same probe.

That's my point, for very module-specific probes (such as the ones ovs module uses or the nf_tables module will use, etc) it's very unlikely another module will use a targeted hook on them.

Reply:
A hash function (a well-known one :-D) is used to for all keys in a hashtable: https://elixir.bootlin.com/linux/latest/source/kernel/bpf/hashtab.c#L653

Reply:
My suggestion is to use a HASH_OF_MAPS with the outer key being the ksym, the inner key being the symbol name and the value being the offset.

Reply:
#99